File: src/main/java/ca/mcgill/science/tepid/server/rest/Jobs.java
package ca.mcgill.science.tepid.server.rest;

import ca.mcgill.science.tepid.common.Destination;
import ca.mcgill.science.tepid.common.PrintJob;
import ca.mcgill.science.tepid.common.Session;
import ca.mcgill.science.tepid.common.ViewResultSet;
import ca.mcgill.science.tepid.common.ViewResultSet.Row;
import ca.mcgill.science.tepid.server.gs.GS;
import ca.mcgill.science.tepid.server.gs.GS.InkCoverage;
import ca.mcgill.science.tepid.server.util.CouchClient;
import ca.mcgill.science.tepid.server.util.QueueManager;
import ca.mcgill.science.tepid.server.util.SessionManager;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.glassfish.jersey.jackson.JacksonFeature;
import org.tukaani.xz.XZInputStream;
import shared.Config;
import shared.ConfigKeys;

import javax.annotation.security.RolesAllowed;
import javax.ws.rs.*;
import javax.ws.rs.client.Client;
import javax.ws.rs.client.ClientBuilder;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.WebTarget;
import javax.ws.rs.container.AsyncResponse;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.Suspended;
import javax.ws.rs.core.*;
import java.io.*;
import java.nio.file.Files;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Path("/jobs")
public class Jobs {
	
	public static final Map<String, Thread> processingThreads = new ConcurrentHashMap<>();
	
	private final WebTarget couchdb = CouchClient.getTepidWebTarget();
	
	private static class JobResultSet extends ViewResultSet<String,PrintJob> {}
	
	@GET
	@Path("/{sam}")
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.APPLICATION_JSON)
	public Collection<PrintJob> listJobs(@PathParam("sam") String sam, @Context ContainerRequestContext req) {
		Session session = (Session) req.getProperty("session");
		if (session.getRole().equals("user") && !session.getUser().shortUser.equals(sam)) {
			return null;
		}
		List<Row<String,PrintJob>> rows = couchdb.path("_design/main/_view").path("byUser").queryParam("key", "\"" + sam + "\"")
		.request(MediaType.APPLICATION_JSON).get(JobResultSet.class).rows;
		Collection<PrintJob> out = new TreeSet<>(new Comparator<PrintJob>() {
			@Override
			public int compare(PrintJob j1, PrintJob j2) {
				Date p1 = j1.getProcessed(), p2 = j2.getProcessed();
				if (j1.getFailed() != null) p1 = j1.started;
				if (j2.getFailed() != null) p2 = j2.started;
				if (p1 == null && p2 == null) return j1.started.compareTo(j2.started);
				if (p1 == null) return -1;
				if (p2 == null) return 1;
				return p2.compareTo(p1) == 0 ? j2.getId().compareTo(j1.getId()) : p2.compareTo(p1);
			}
		});
		for (Row<String,PrintJob> r : rows) {
			out.add(r.value);
		}
		return out;
	}

	@POST
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.APPLICATION_JSON)
	@Consumes(MediaType.APPLICATION_JSON)
	public String newJob(PrintJob j, @Context ContainerRequestContext req) {
		j.setUserIdentification(((Session) req.getProperty("session")).getUser().shortUser);
		j.setDeleteDataOn(System.currentTimeMillis() + SessionManager.getInstance().queryUserCache(j.getUserIdentification()).jobExpiration);
		System.out.println(j);
		return couchdb.request(MediaType.TEXT_PLAIN).post(Entity.entity(j, MediaType.APPLICATION_JSON)).readEntity(String.class);
	}

	@PUT
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.TEXT_PLAIN)
	@Path("/{id}")
	public String addJobData(InputStream is, @PathParam("id") final String id) {
		System.out.println(id + " Receiving job data");
		File tmpDir = new File(System.getProperty("os.name").startsWith("Windows") ? System.getProperty("java.io.tmpdir") + "\\tepid" : "/tmp/tepid");
		if (!tmpDir.exists()) {
			tmpDir.mkdirs();
		}
		try {
			//write compressed job to disk
			final File tmpXz = new File(tmpDir.getAbsolutePath() + "/" + id +  ".ps.xz");
			Files.copy(is, tmpXz.toPath());
			is.close();
			//let db know we have received data
			PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
			j.setFile(tmpXz.getAbsolutePath());
			j.setReceived(new Date());
			couchdb.path(id).request(MediaType.TEXT_PLAIN).put(Entity.entity(j, MediaType.APPLICATION_JSON));
			Thread processing = new Thread("Job Processing for " + id) {
				@Override
				public void run() {
					try {
						//decompress data
						File tmp = File.createTempFile("tepid", ".ps");
						tmp.delete();
						XZInputStream decompress = new XZInputStream(new FileInputStream(tmpXz));
						Files.copy(decompress, tmp.toPath());
						//count pages
						List<InkCoverage> inkCov = GS.inkCoverage(tmp);
						int color = 0;
						for (InkCoverage i : inkCov) { 
							if (!i.monochrome) color++;
						}
						//update page count and status in db
						PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
						j.setPages(inkCov.size());
						j.setColorPages(color);
						j.setProcessed(new Date());
						System.err.println(id + " setting stats ("+inkCov.size()+" pages, "+color+" color)");
						couchdb.path(id).request(MediaType.TEXT_PLAIN).put(Entity.entity(j, MediaType.APPLICATION_JSON));
						//check if user has color printing enabled
						if (color > 0 && !SessionManager.getInstance().queryUser(j.getUserIdentification(), null).colorPrinting) {
							failJob(id, "Color disabled");
						} else {
							//check if user has sufficient quota to print this job
							if (new Users().getQuota(j.getUserIdentification()) < inkCov.size() - color + color * 3) {
								failJob(id, "Insufficient quota");
							} else {
								//add job to the queue
								j = QueueManager.assignDestination(id);
								Destination dest = couchdb.path(j.getDestination()).request(MediaType.APPLICATION_JSON).get(Destination.class);
								if (sendToSMB(tmp, dest)) {
									j.setPrinted(new Date());
									couchdb.path(id).request(MediaType.TEXT_PLAIN).put(Entity.entity(j, MediaType.APPLICATION_JSON));
									System.err.println(j.getId() + " sent to destination");
								} else {
									failJob(id, "Could not send to destination");
								}
							}
						}
						tmp.delete();
					} catch (Exception e) {
						e.printStackTrace();
						failJob(id, "Exception during processing");
					}
					processingThreads.remove(id);
				}
			};
			processingThreads.put(id, processing);
			processing.start();
			return "Job data for " + id + " successfully uploaded";
		} catch (IOException e) {
			e.printStackTrace();
			failJob(id, " Exception during upload");
		}
		

		return "Job data upload for " + id + " FAILED";
	}
	
	public void failJob(String id, String error) {
		PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
		j.setFailed(new Date(), error);
		couchdb.path(j.getId()).request(MediaType.TEXT_PLAIN).put(Entity.entity(j, MediaType.APPLICATION_JSON));
	}
	
	public static boolean sendToSMB(File f, Destination destination) {
		if (destination.getPath() == null || destination.getPath().trim().isEmpty()) {
			//this is a dummy destination
			try {
				Thread.sleep(4000);
			} catch (InterruptedException e) {
			}
			return true;
		}
		System.setProperty("jcifs.smb.client.useNTSmbs", "false");
		try {
			Process p = new ProcessBuilder("smbclient", "//" + destination.getPath(), destination.getPassword(), "-c",
			"print " + f.getAbsolutePath(), "-U", destination.getDomain() + "\\" + destination.getUsername()).start();
			p.waitFor();			
		} catch (IOException | InterruptedException e) {
			return false;
		}
		return true;
	}
	
//	public static boolean sendToSMB(File f, Destination destination) {
//		if (destination.getPath() == null || destination.getPath().trim().isEmpty()) {
//			//this is a dummy destination
//			try {
//				Thread.sleep(4000);
//			} catch (InterruptedException e) {
//			}
//			return true;
//		}
//		System.setProperty("jcifs.smb.client.useNTSmbs", "false");
//		try {
//			String userName = destination.getUsername(), password = destination.getPassword();
//			SmbFileOutputStream sfos = new SmbFileOutputStream(new SmbFile("smb://" + destination.getPath(), new NtlmPasswordAuthentication(destination.getDomain(), userName, password)));
//			Files.copy(f.toPath(), sfos);
//			sfos.close();
//		} catch (IOException e) {
//			return false;
//		}
//		return true;
//	}
	
	@GET
	@Path("/job/{id}/_changes")
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.APPLICATION_JSON)
	public void getChanges(@PathParam("id") String id, @Context UriInfo uriInfo, @Suspended AsyncResponse ar, @Context ContainerRequestContext req) {
		PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
		Session session = (Session) req.getProperty("session");
		if (session.getRole().equals("user") && !session.getUser().shortUser.equals(j.getUserIdentification())) {
			ar.resume(Response.status(Response.Status.UNAUTHORIZED).entity("You cannot access this resource").type(MediaType.TEXT_PLAIN).build());
		}
		WebTarget target = couchdb.path("_changes").queryParam("filter", "main/byJob").queryParam("job", id);
		MultivaluedMap<String, String> qp = uriInfo.getQueryParameters();
		if (qp.containsKey("feed")) target = target.queryParam("feed", qp.getFirst("feed"));
		if (qp.containsKey("since")) target = target.queryParam("since", qp.getFirst("since"));
		//TODO find a way to make this truly asynchronous
		String changes = target.request().get(String.class);
		if (!ar.isDone() && !ar.isCancelled()) {
			ar.resume(changes);
		}
	}
	
	@GET
	@Path("/job/{id}")
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.APPLICATION_JSON)
	public Response getJob(@PathParam("id") String id, @Context UriInfo uriInfo, @Context ContainerRequestContext req) {
		PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
		Session session = (Session) req.getProperty("session");
		if (session.getRole().equals("user") && !session.getUser().shortUser.equals(j.getUserIdentification())) {
			return Response.status(Response.Status.UNAUTHORIZED).entity("You cannot access this resource").type(MediaType.TEXT_PLAIN).build();
		}
		return Response.ok(j).build();
	}
	
	@PUT
	@Path("/job/{id}/refunded")
	@RolesAllowed({"ctfer", "elder"})
	@Produces(MediaType.APPLICATION_JSON)
	public Response setJobRefunded(@PathParam("id") String id, @Context ContainerRequestContext req, boolean refunded) {
		PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
		Session session = (Session) req.getProperty("session");
		if (session.getRole().equals("ctfer") && session.getUser().shortUser.equals(j.getUserIdentification())) {
			return Response.status(Response.Status.UNAUTHORIZED).entity("You cannot refund your own jobs").type(MediaType.TEXT_PLAIN).build();
		}
		j.setRefunded(refunded);
		couchdb.path(j.getId()).request().put(Entity.entity(j, MediaType.APPLICATION_JSON));
		return Response.ok().build();
	}
	
	@POST
	@Path("/job/{id}/reprint")
	@RolesAllowed({"user", "ctfer", "elder"})
	@Produces(MediaType.TEXT_PLAIN)
	public Response reprintJob(@PathParam("id") String id, @Context ContainerRequestContext req) {
		final PrintJob j = couchdb.path(id).request(MediaType.APPLICATION_JSON).get(PrintJob.class);
		if (j.getFile() == null || !new File(j.getFile()).exists()) {
			return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Data for this job no longer exists").type(MediaType.TEXT_PLAIN).build();
		}
		Session session = (Session) req.getProperty("session");
		if (session.getRole().equals("user") && !session.getUser().shortUser.equals(j.getUserIdentification())) {
			return Response.status(Response.Status.UNAUTHORIZED).entity("You cannot reprint someone else's job").type(MediaType.TEXT_PLAIN).build();
		}
		PrintJob reprint = new PrintJob();
		reprint.setName(j.getName());
		reprint.setOriginalHost("REPRINT");
		reprint.setQueueName(j.getQueueName());
		reprint.setUserIdentification(j.getUserIdentification());
		reprint.setDeleteDataOn(System.currentTimeMillis() + SessionManager.getInstance().queryUserCache(j.getUserIdentification()).jobExpiration);
		System.out.println(reprint);
		final String newId = couchdb.request(MediaType.APPLICATION_JSON).post(Entity.entity(reprint, MediaType.APPLICATION_JSON)).readEntity(ObjectNode.class).get("id").asText();
		new Thread("Reprint " + id) {
			@Override
			public void run() {
				try {
					addJobData(new FileInputStream(new File(j.getFile())), newId);
				} catch (FileNotFoundException e) {
				}
			}
		}.start();
		return Response.ok("Reprinted " + id + " new id " + newId).build();
	}

}

File: src/main/java/ca/mcgill/science/tepid/server/rest/Jobs.kt
package ca.mcgill.science.tepid.server.rest

import ca.mcgill.science.tepid.models.bindings.CTFER
import ca.mcgill.science.tepid.models.bindings.ELDER
import ca.mcgill.science.tepid.models.bindings.USER
import ca.mcgill.science.tepid.models.data.PrintJob
import ca.mcgill.science.tepid.models.data.PutResponse
import ca.mcgill.science.tepid.server.auth.AuthenticationManager
import ca.mcgill.science.tepid.server.db.DB
import ca.mcgill.science.tepid.server.db.remapExceptions
import ca.mcgill.science.tepid.server.printing.Printer
import ca.mcgill.science.tepid.server.util.Utils
import ca.mcgill.science.tepid.server.util.failBadRequest
import ca.mcgill.science.tepid.server.util.failInternal
import ca.mcgill.science.tepid.server.util.failUnauthorized
import ca.mcgill.science.tepid.server.util.getSession
import ca.mcgill.science.tepid.server.util.logMessage
import org.apache.logging.log4j.kotlin.Logging
import java.io.FileInputStream
import java.io.InputStream
import java.util.concurrent.TimeUnit
import javax.annotation.security.RolesAllowed
import javax.ws.rs.Consumes
import javax.ws.rs.GET
import javax.ws.rs.POST
import javax.ws.rs.PUT
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.ws.rs.container.ContainerRequestContext
import javax.ws.rs.core.Context
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.UriInfo

@Path("/jobs")
class Jobs {

    @POST
    @RolesAllowed(USER, CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun newJob(j: PrintJob, @Context ctx: ContainerRequestContext): PutResponse {
        val session = ctx.getSession()
        remapExceptions { DB.queues.read(j.queueId) }
        j.userIdentification = session.user.shortUser
        j.deleteDataOn = j.getJobExpiration()
        logger.debug(logMessage("starting new print job", "name" to j.name, "for" to session.user.longUser))
        return remapExceptions { DB.printJobs.create(j) }
    }

    @GET
    @Path("/{id}")
    @RolesAllowed(USER, CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    fun getJob(@PathParam("id") id: String, @Context uriInfo: UriInfo, @Context ctx: ContainerRequestContext): PrintJob {
        val session = ctx.getSession()
        logger.trace(logMessage("queried job", "id" to id))
        val j = DB.printJobs.read(id)
        if (session.role == USER && session.user.shortUser != j.userIdentification)
            failUnauthorized("You cannot access this resource")
        return j
    }

    @PUT
    @RolesAllowed(USER, CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    @Path("/{id}/data")
    fun addJobData(input: InputStream, @PathParam("id") id: String): PutResponse {
        logger.debug(logMessage("receiving job data", "id" to id))
        val (success, message) = Printer.print(id, input)
        if (!success)
            failBadRequest(message)
        return PutResponse(true, id, "")
    }

    @PUT
    @Path("/{id}/refunded")
    @RolesAllowed(CTFER, ELDER)
    @Produces(MediaType.APPLICATION_JSON)
    fun setJobRefunded(@PathParam("id") id: String, refunded: Boolean): PutResponse {
        val result = DB.printJobs.update(id) {
            isRefunded = refunded
            logger.debug(logMessage("refunded job", "id" to id))
        }
        return PutResponse(result.isRefunded == refunded, result.getId(), result.getRev())
    }

    @POST
    @Path("/{id}/reprint")
    @RolesAllowed(USER, CTFER, ELDER)
    @Produces(MediaType.TEXT_PLAIN)
    fun reprintJob(@PathParam("id") id: String, @Context ctx: ContainerRequestContext): String {
        val session = ctx.getSession()
        val j = DB.printJobs.read(id)
        val file = Utils.existingFile(j.file) ?: failInternal("Data for this job no longer exists")
        if (session.role == USER && session.user.shortUser != j.userIdentification)
            failUnauthorized("You cannot reprint someone else's job")
        val reprint = PrintJob(
            name = j.name,
            originalHost = "REPRINT",
            queueId = j.queueId,
            userIdentification = j.userIdentification,
            deleteDataOn = j.getJobExpiration()
        )
        logger.debug(logMessage("reprinted", "name" to reprint.name))
        val response = remapExceptions { DB.printJobs.create(reprint) }
        val newId = response.id
        Utils.startCaughtThread("Reprint $id", logger) {
            val (success, message) = Printer.print(newId, FileInputStream(file))
            if (!success)
                logger.error { "Failed to reprint job: $message" }
        }
        logger.debug(logMessage("reprinted job", "id" to id, "newId" to newId))
        return "Reprinted $id, new id $newId"
    }

    private fun PrintJob.getJobExpiration(): Long {
        val userId = userIdentification ?: return 0
        return System.currentTimeMillis() + (AuthenticationManager.queryUserDb(userId)?.jobExpiration
            ?: TimeUnit.DAYS.toMillis(7))
    }

    private companion object : Logging
}
